Api routes refactor

BenMorganIO opened the pull request.

BenMorganIO committed:
use multiple args for resources

This is a nice refactor. I’ve also gone and moved the resources with no
blocks to the top. This way I can immediately see where I can refactor
using multiple args instead of recalling the resource method.

Conflicts:
	api/config/routes.rb

BenMorganIO committed:
put the puts in a single put

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://benmorganio-api-routes-refactor.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://benmorganio-api-routes-refactor.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

JDutil commented on api/config/routes.rb:
```diff
@@ -1,56 +1,40 @@
-        put :generate_api_key
-        put :clear_api_key
-      end
+      put :generate_api_key, :clear_api_key, on: :member
```
Personally I prefer these remain as blocks for readability or perhaps:
```
put :clear_api_key,
    :generate_api_key,
    on: :member
```

BenMorganIO replied:
The single `put` method should speed up ruby by 1 nanosecond since there's less methods being called, so I could argue performance, but that wouldn't help.

I like how the blocks, being removed, free up lines of code, but I also like how you have organized your `put` method. I'll opt for that and get best of both worlds.

JDutil commented:
I'm a fan of alphabetization for readability, and if we're going to refactor this to clean it up would be nice to alphabetize as well.

BenMorganIO committed:
list the puts for readability

BenMorganIO committed:
reorganize alphabetically and semantically

get methods, then put methods, then concerns, then resources. Resources
are also listed via their complexity as well. No blocks at top
organized by their `only` arrays then theirs the resources with the
blocks below them.

BenMorganIO commented on api/config/routes.rb:
```diff
@@ -1,127 +1,109 @@
   namespace :api, defaults: { format: 'json' } do
-    resources :promotions, only: [:show]
+    get '/config/money', to: 'config#money'
+    get '/config', to: 'config#show'
```
This should probably become a `resource`.

BenMorganIO committed:
use a resource instead

Resource should come before resources.

gmacdougall commented:
Looks good to me.

JDutil commented:
Just a couple specs to fix:

```
Failures:

  1) Spree::Api::ConfigController returns Spree::Money settings
     Failure/Error: api_get :money
     ActionController::UrlGenerationError:
       No route matches {:action=>"money", :controller=>"spree/api/config", :format=>:json}
     # ./spec/support/controller_hacks.rb:27:in `api_process'
     # ./spec/support/controller_hacks.rb:10:in `api_get'
     # ./spec/controllers/spree/api/config_controller_spec.rb:12:in `block (2 levels) in <module:Spree>'
     # /home/ubuntu/spree/vendor/bundle/ruby/2.2.0/bin/rspec:23:in `load'
     # /home/ubuntu/spree/vendor/bundle/ruby/2.2.0/bin/rspec:23:in `<main>'

  2) Spree::Api::ConfigController returns some configuration settings
     Failure/Error: api_get :show
     ActionController::UrlGenerationError:
       No route matches {:action=>"show", :controller=>"spree/api/config", :format=>:json}
     # ./spec/support/controller_hacks.rb:27:in `api_process'
     # ./spec/support/controller_hacks.rb:10:in `api_get'
     # ./spec/controllers/spree/api/config_controller_spec.rb:18:in `block (2 levels) in <module:Spree>'
     # /home/ubuntu/spree/vendor/bundle/ruby/2.2.0/bin/rspec:23:in `load'
     # /home/ubuntu/spree/vendor/bundle/ruby/2.2.0/bin/rspec:23:in `<main>'

Finished in 1 minute 38.51 seconds (files took 3.76 seconds to load)
444 examples, 2 failures, 5 pending

Failed examples:

rspec ./spec/controllers/spree/api/config_controller_spec.rb:11 # Spree::Api::ConfigController returns Spree::Money settings
rspec ./spec/controllers/spree/api/config_controller_spec.rb:17 # Spree::Api::ConfigController returns some configuration settings
```

BenMorganIO commented:
That's probably because its looking for configs and not config.

BenMorganIO committed:
set controller to config instead of configs

BenMorganIO commented:
I checked inside of the `Spree::Api::ConfigController` and I didn't see a `#show` action. Does anyone know where that's defined?

BenMorganIO commented:
There's also no `#money` action...

JDutil commented:
When I look at the api "config" controller it looks like it does nothing to me...  Does it actually serve a purpose or was it introduced, and never actually finished?

BenMorganIO commented:
@JDutil I'm actually more concerned as to why the routes are passing.

BenMorganIO commented:
@JDutil I just went and made #6094.

JDutil commented:
This should be rebased with master

BenMorganIO committed:
fix merge conflicts
